### PR TITLE
[Impeller] refactor clip rendering.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Now that we don't have the entity pass structure, clips can be treated as a kind of structured data and not a generic entity. Doing so lets us simplify some of the clip management and remove some usage of raw pointers. This change also removes the ClipRestoreContents, which are only ever created and then immediately rendered - we can instead replace this with a function.

The ClipCoverageStack now has explicit Clip/Restore methods that internally apply the same logic, but without involving the entity object.

Redo of https://github.com/flutter/engine/pull/55784

### CODE REVIEW

**Review by bdero — APPROVED** [review]
LGTM!

File: impeller/display_list/canvas.cc [hunk 1]
```diff
@@ -550,34 +550,84 @@ void Canvas::DrawCircle(const Point& center,
...
+                                      GetGlobalPassPosition(),  //
+                                      clip_depth,               //
+                                      GetClipHeightFloor()      //
+      );
+
+  if (clip_state_result.clip_did_change) {
+    // We only need to update the pass scissor if the clip state has changed.
+    SetClipScissor(clip_coverage_stack_.CurrentClipCoverage(),
+                   *render_passes_.back().inline_pass_context->GetRenderPass(),
+                   GetGlobalPassPosition());
```
[the remark this answers is not in the document]
Comment (bdero) [reply impeller/display_list/canvas.cc:601]: Eh rounded rect clips are pretty common too. So maybe not _most_ clips, but...

[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/display_list/canvas.cc:601]: a lot of the "to be safe" clips are clip rects set to hardEdge

Actually, we could _honor_ the hardEdge setting for axis aligned and discard the factional coverage?

File: impeller/display_list/canvas.cc [hunk 2]
```diff
@@ -550,34 +550,84 @@ void Canvas::DrawCircle(const Point& center,
...
+
+  std::optional<Rect> clip_coverage = geometry.GetCoverage(clip_transform);
+  if (!clip_coverage.has_value()) {
+    return;
+  }
+
+  ClipContents clip_contents(
+      clip_coverage.value(),
+      geometry.IsAxisAlignedRect() &&
+          GetCurrentTransform().IsTranslationScaleOnly());
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/display_list/canvas.cc]: Done

File: impeller/display_list/canvas.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (bdero) [reply impeller/display_list/canvas.cc:601]: > There are a lot of devices with something like a 2.65 DPI, but for the integrer variants yeah. Will do in follow up?

Yeah of course, just thinking aloud. No need to do it in this patch.

> a lot of the "to be safe" clips are clip rects set to hardEdge
> 
> Actually, we could _honor_ the hardEdge setting for axis aligned and discard the factional coverage?

That seems like a good idea. From the descriptions in [here](https://docs.flutter.dev/release/breaking-changes/clip-behavior) and [here](https://api.flutter.dev/flutter/dart-ui/Clip.html), it sounds like the clip should always round inwards on all sides to intentionally avoid anti-aliasing on long boxy edges?

[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/display_list/canvas.cc:601]: not sure if round in or round out. The SkCanvas API just has a `bool doAntiAlias`

**Pull request merged**